analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarCompilationStartAnalysisContext.cs

```diff
@@ -31,9 +31,6 @@ public sealed class SonarCompilationStartAnalysisContext : SonarAnalysisContextB
     public void RegisterSymbolAction(Action<SonarSymbolReportingContext> action, params SymbolKind[] symbolKinds) =>
         Context.RegisterSymbolAction(x => action(new(AnalysisContext, x)), symbolKinds);
 
-    public void RegisterSemanticModelAction(Action<SonarSematicModelReportingContext> action) =>
-        Context.RegisterSemanticModelAction(x => action(new(AnalysisContext, x)));
```

Comment:
The point in time given by RegisterSemanticModelAction is still right. The binding phase needs to be done, before you can successfully [query the semantic model for symbols](https://github.com/dotnet/roslyn/blob/main/docs/wiki/Roslyn-Overview.md#compiler-pipeline-functional-areas).

> First the parse phase, where source is tokenized and parsed into syntax that follows the language grammar. Second the declaration phase, where declarations from source and imported metadata are analyzed to form named symbols. Next the bind phase, where identifiers in the code are matched to symbols.

We need these compiler pipeline entry points. Please do not remove any (we should instead add more). The right thing to do is to ignore the sematic model passed in `RegisterSemanticModelAction` and create a new one with `Compilation.GetSemanticModel(syntaxTree)` in the action instead.

Reply:
We need to revert the changes fully, as Only using a different semantic model will not fix the regression.
Calling `ShouldGenerateMetrics` from the `SemanticModelAction` is also slowing down the analysis significantly because it uses two concurrent collections inside.
The decision is to revert as close as possible to the original state in order to finish the sprint and make it to the next SQ release.

Making the change to `SemanticModelAction` makes sense, and at the same time it should be done as part of a proper sprint that focuses on it, so we can come up with the best-performing solution that is backed by a proper and thorough validation.

Reply:
@martin-strecker-sonarsource  I agree with @csaba-sagi-sonarsource  - one thing I learned from this sprint is that even if the change seems to make absolute sense in code and in theory we still need to do a thorough performance validation as it might behave in an unexpected behavior.

From my point of view - I would add an issue for this and I'd tackle it in the next hardening (as this one is coming to an end tomorrow).

I'll also add a card to do a sprint dedicated to the hardening of the UtilityAnalyzers (specified and planned as this one had only scope to fix the regression). I think we can bring performance and cleanup value this way.